src/main/java/io/fabric8/maven/docker/config/handler/property/PropertyConfigHandler.java

```diff
@@ -41,165 +40,167 @@ public String getType() {
+        return valueProvider.getString(FROM, config == null ? null : config.getFrom()) != null ||
+                valueProvider.getMap(FROM_EXT, config == null ? null : config.getFromExt()) != null ||
+                valueProvider.getString(DOCKER_FILE, config == null || config.getDockerFile() == null ? null : config.getDockerFile().getAbsolutePath()) != null ||
+                valueProvider.getString(DOCKER_FILE_DIR, config == null || config.getDockerArchive() == null ? null : config.getDockerArchive().getAbsolutePath()) != null;
     }
 
 
-    private BuildImageConfiguration extractBuildConfiguration(String prefix, Properties properties) {
-        if (!buildConfigured(prefix, properties)) {
+    private BuildImageConfiguration extractBuildConfiguration(BuildImageConfiguration config, ValueProvider valueProvider) {
+        if (!buildConfigured(config, valueProvider)) {
             return null;
         }
+
         return new BuildImageConfiguration.Builder()
-                .cmd(withPrefix(prefix, CMD, properties))
-                .cleanup(withPrefix(prefix, CLEANUP, properties))
-                .nocache(withPrefix(prefix, NOCACHE, properties))
-                .optimise(withPrefix(prefix, OPTIMISE, properties))
-                .entryPoint(withPrefix(prefix, ENTRYPOINT, properties))
-                .assembly(extractAssembly(prefix, properties))
-                .env(mapWithPrefix(prefix, ENV, properties))
-                .args(mapWithPrefix(prefix, ARGS, properties))
-                .labels(mapWithPrefix(prefix,LABELS,properties))
-                .ports(extractPortValues(prefix, properties))
-                .runCmds(extractRunCommands(prefix,properties))
-                .from(withPrefix(prefix, FROM, properties))
-                .fromExt(mapWithPrefix(prefix,FROM_EXT,properties))
-                .registry(withPrefix(prefix, REGISTRY, properties))
-                .volumes(listWithPrefix(prefix, VOLUMES, properties))
-                .tags(listWithPrefix(prefix, TAGS, properties))
-                .maintainer(withPrefix(prefix, MAINTAINER, properties))
-                .workdir(withPrefix(prefix, WORKDIR, properties))
-                .skip(withPrefix(prefix, SKIP_BUILD, properties))
-                .imagePullPolicy(withPrefix(prefix, IMAGE_PULL_POLICY_BUILD, properties))
-                .dockerArchive(withPrefix(prefix, DOCKER_ARCHIVE, properties))
-                .buildOptions(mapWithPrefix(prefix, BUILD_OPTIONS, properties))
-                .dockerFile(withPrefix(prefix, DOCKER_FILE, properties))
-                .dockerFileDir(withPrefix(prefix, DOCKER_FILE_DIR, properties))
-                .filter(withPrefix(prefix, FILTER, properties))
-                .user(withPrefix(prefix, USER, properties))
-                .healthCheck(extractHealthCheck(prefix, properties))
+                .cmd(valueProvider.getString(CMD, config == null ? null : config.getCmd()))
```

Comment:
Hm, thought about alternatives to that. That is one way, but then we have to repeat the full Builder structure, right?

I guess Java8 and Optional with lambdas isn't an option?

Reply:
Why would we need to rebuild the builder structure ? Don't really see this (as the members of the builder is already initialized with null).

d-m-p started as a Java 7 project at that time and hence doesn't use optionals. But I think in our case it would help quite a bit, so if you have a good suggestion, we I wouldn't mind to introduce Java 8 features.

Reply:
Just extracting config != null check would look something like this:
```
if(config != null) {
    return new RunImageConfiguration.Builder()
        .capAdd(valueProvider.getList(CAP_ADD, config.getCapAdd()))
        .capDrop(valueProvider.getList(CAP_DROP, config.getCapDrop()))
        .securityOpts(valueProvider.getList(SECURITY_OPTS, config.getSecurityOpts()))
        .cmd(extractArguments(valueProvider, CMD, config.getCmd()))
        .....
        .build();
}else{
    return new RunImageConfiguration.Builder()
        .capAdd(valueProvider.getList(CAP_ADD, null))
        .capDrop(valueProvider.getList(CAP_DROP, null))
        .securityOpts(valueProvider.getList(SECURITY_OPTS, null))
        .cmd(extractArguments(valueProvider, CMD, null))
        .....
        .build();
}
```

A bit cleaner to see,, but easier to miss when adding new properties.

With java8 and lambdas, it could be something like this (Optionals or local fn):
```
static <T,R> R orNull(T config, java.util.function.Function<T,R> fn) {
    if(config != null) {
       return fn.apply(config);
    }
    return null;
}
....
return new RunImageConfiguration.Builder()
    .capAdd(valueProvider.getList(CAP_ADD, orNull(config, RunImageConfiguration::getCapAdd)))
    .capDrop(valueProvider.getList(CAP_DROP, Optional.of(config).map(RunImageConfiguration::getCapDrop).orElse(null)))
    ...
    .build();
```

Not sure if the RunImageConfiguration:: can somehow be interfered from T.. But a lot cleaner.

Anyway, adding Java8 as a requirement would hopefully not be an issue for users of the plugin, last public update of Java 7 was 2015 it seems?... At least we're not stuck in Java 6.. :) 

**BUT**,  I've seen far too many cases when Java version updates irritates, and I really really would not want to wait for any such disputes to clear up before a new release is made.. So unless you insist on updating, and still having a pretty quick release, I prefer ugly over Java 8.

Reply:
* writing comment to not hide as outdated*

Reply:
Ok, lets keep it that way and switch over to Optional soon after. I don't mind to set Java 8 as minimal requirement, and we are not yet 1.0.0 so should be easy todo.

And you are right Java 7 should be dead now, especially for build tools.